helix-core/src/main/java/org/apache/helix/manager/zk/ZKHelixManager.java

```diff
@@ -1285,4 +1266,85 @@ public void handleSessionEstablishmentError(Throwable error) throws Exception {
   public Long getSessionStartTime() {
     return _sessionStartTime;
   }
+
+  private RealmAwareZkClient buildRealmAwareZkClient() {
+    final String shardingKey = buildShardingKey();
+    PathBasedZkSerializer zkSerializer =
+        ChainedPathZkSerializer.builder(new ZNRecordSerializer()).build();
+
+    RealmAwareZkClient.RealmAwareZkConnectionConfig connectionConfig =
+        new RealmAwareZkClient.RealmAwareZkConnectionConfig.Builder()
+            .setRealmMode(RealmAwareZkClient.RealmMode.SINGLE_REALM)
+            .setZkRealmShardingKey(shardingKey)
+            .setSessionTimeout(_sessionTimeout).build();
+
+    RealmAwareZkClient.RealmAwareZkClientConfig clientConfig =
+        new RealmAwareZkClient.RealmAwareZkClientConfig();
+
+    clientConfig.setZkSerializer(zkSerializer)
+        .setConnectInitTimeout(_connectionInitTimeout)
+        .setMonitorType(_instanceType.name())
+        .setMonitorKey(_clusterName)
+        .setMonitorInstanceName(_instanceName)
+        .setMonitorRootPathOnly(isMonitorRootPathOnly());
+
+    RealmAwareZkClient newClient;
+    if (_instanceType == InstanceType.ADMINISTRATOR) {
+      newClient = buildSharedZkClient(connectionConfig, clientConfig);
+    } else {
+      newClient = buildDedicatedZkClient(connectionConfig, clientConfig);
+    }
+
+    return newClient;
+  }
+
+  private RealmAwareZkClient buildSharedZkClient(
```

Comment:
Is it possible to combine `buildSharedZkClient` and `buildDedicatedZkClient` to reduce duplication? Both instances of the factories extend `HelixZkClientFactory`.

Reply:
I tried. We have 2 options:
1. As you said, we can pass in the factory as a parameter. But we have to make `HelixZkClientFactory` public. Shall we do this?
`public abstract class HelixZkClientFactory implements RealmAwareZkClientFactory`

2. It is the implemented as I do. Keep`HelixZkClientFactory` as it is. I prefer this considering that we don't want HelixZkClientFactory public.